File: Crowmask.DomainModeling/ActivityStreamsIdMapper.fs
namespace Crowmask.DomainModeling

open System

type ActivityStreamsIdMapper(host: ICrowmaskHost) =
    member _.ActorId =
        $"https://{host.Hostname}/api/actor"

    member _.GetTransientId() =
        $"https://{host.Hostname}#transient-{Guid.NewGuid()}"

    member _.GetObjectType(upstream_type) =
        match upstream_type with
        | UpstreamSubmission _ -> "Note"
        | UpstreamJournal _ -> "Article"

    member _.GetObjectId(upstream_type) =
        match upstream_type with
        | UpstreamSubmission submitid -> $"https://{host.Hostname}/api/submissions/{submitid}"
        | UpstreamJournal journalid -> $"https://{host.Hostname}/api/journals/{journalid}"

    member this.GetObjectId(upstream_type: UpstreamType, interaction: Interaction) =
        $"{this.GetObjectId(upstream_type)}/interactions/{interaction.Id}/notification"

File: Crowmask.IdMapping/ActivityStreamsIdMapper.cs
﻿using Crowmask.DomainModeling;

namespace Crowmask.IdMapping
{
    public class ActivityStreamsIdMapper(ICrowmaskHost crowmaskHost)
    {
        public string ActorId =>
            $"https://{crowmaskHost.Hostname}/api/actor";

        public string GetTransientId() =>
            $"{ActorId}#transient-{Guid.NewGuid()}";

        public string GetObjectType(JointIdentifier identifier) =>
            identifier is JointIdentifier.SubmissionIdentifier ? "Note"
            : identifier is JointIdentifier.JournalIdentifier ? "Article"
            : throw new NotImplementedException();

        public string GetObjectId(JointIdentifier identifier) =>
            identifier is JointIdentifier.SubmissionIdentifier s
                ? $"https://{crowmaskHost.Hostname}/api/submissions/{s.submitid}"
            : identifier is JointIdentifier.JournalIdentifier j
                ? $"https://{crowmaskHost.Hostname}/api/journals/{j.journalid}"
            : throw new NotImplementedException();

        public JointIdentifier? GetJointIdentifier(string objectId)
        {
            if (!Uri.TryCreate(objectId, UriKind.Absolute, out Uri? uri))
                return null;

            if (!int.TryParse(uri.AbsolutePath.Split('/').Last(), out int id))
                return null;

            foreach (var candidate in new[]
            {
                JointIdentifier.NewSubmissionIdentifier(id),
                JointIdentifier.NewJournalIdentifier(id)
            })
            {
                if (GetObjectId(candidate) == objectId)
                    return candidate;
            }

            return null;
        }

        public string GetObjectId(JointIdentifier identifier, Interaction interaction) =>
            $"{GetObjectId(identifier)}/interactions/{interaction.Id}/notification";
    }
}
